File: Python/programmers/etc/level1/공원_산책.py
move = {"E":(0,1),"W":(0,-1),"S":(1,0),"N":(-1,0)}

def solution(park, routes):
    x,y,N,M = 0, 0, len(park), len(park[0])
    for i, row in enumerate(park):
        if "S" in row:
            x, y = i, row.find("S")
            break

    for route in routes:
        dx, dy = move[route[0]]
        nx, ny = x,y
        for _ in range(int(route[2])): 
            if 0 <= nx+dx < N and 0 <= ny+dy < M and park[nx+dx][ny+dy] != "X":
                nx,ny = nx+dx, ny+dy
            else:
                nx,ny = x,y
                break
        x,y = nx,ny

    return [x,y]


park = ["SOO","OOO","OOO"]
routes = ["E 2","S 2","W 1"]
print(solution(park, routes)) # [2,1]

park = ["SOO","OXX","OOO"]
routes = ["E 2","S 2","W 1"]
print(solution(park, routes)) # [0,1]

park = ["OSO","OOO","OXO","OOO"]
routes = ["E 2","S 3","W 1"]
print(solution(park, routes)) # [0,0]

File: Java/programmers/etc/level1/공원산책.java
package CodeTest.Java.programmers.etc.level1;

public class 공원산책 {
    public int[] solution(String[] park, String[] routes) {
        int N = park.length;
        int M = park[0].length();
        
        int x = 0;
        int y = 0;
        
        for (int i=0; i<N; i++) {
            if (park[i].contains("S")) {
                x = i;
                y = park[i].indexOf("S");
            }
        }
        
        for (String route : routes) {
            String d = route.split(" ")[0];
            int move = Integer.parseInt(route.split(" ")[1]);
            
            int nx = x;
            int ny = y;
            
            for (int i=0; i<move; i++) {
                switch(d) {
                    case "S" : nx++; break;
                    case "N" : nx--; break;
                    case "E" : ny++; break;
                    case "W" : ny--; break;
                }
                if (nx >= 0 && nx < N && ny >= 0 && ny < M) {
                    if (park[nx].substring(ny, ny+1).equals("X")) break;
                
                    if (i == move-1) {
                        x = nx;
                        y = ny;
                    }
                }
            }
        }
        
        return new int[]{x, y};
    }
}
